src/operators/batch.rs

```diff
@@ -0,0 +1,92 @@
+use std::{
+    task::Poll,
+    time::{Duration, Instant},
+};
+
+use pyo3::{Python, ToPyObject};
+
+use crate::{pyo3_extensions::TdPyAny, unwrap_any};
+
+use super::stateful_unary::{LogicFate, StatefulLogic};
+
+pub(crate) struct BatchLogic {
+    size: usize,
+    timeout: Duration,
+    last_drain: Instant,
+    acc: Vec<TdPyAny>,
+}
+
+impl BatchLogic {
+    pub(crate) fn builder(size: usize, timeout: Duration) -> impl Fn(Option<TdPyAny>) -> Self {
+        move |resume_snapshot| {
+            let acc = resume_snapshot
+                .and_then(|state| -> Option<Vec<TdPyAny>> {
+                    unwrap_any!(Python::with_gil(|py| state.extract(py)))
+                })
+                .unwrap_or_default();
+            Self {
+                size,
+                acc,
+                timeout,
+                last_drain: Instant::now(),
+            }
+        }
+    }
+
+    /// Drain self.acc, convert it to a TdPyAny and return it.
+    /// If self.acc is empty, return None, but still set self.last_drain.
+    fn drain_acc(&mut self) -> Option<TdPyAny> {
+        self.last_drain = Instant::now();
+        if self.acc.is_empty() {
+            None
+        } else {
+            Some(Python::with_gil(|py| {
+                self.acc
+                    .drain(..)
+                    .collect::<Vec<TdPyAny>>()
```

Comment:
My idea here was to avoid repeated allocations for the `self.acc` vec inside `BatchLogic`. Since the batch size is fixed, keeping the same vec around means it only allocates `n` times to get enough space once, rather than at each batch for the same key. But now that you point it out, the logic is not reused between batches right? So yes, it makes sense to use `std::mem::take` here, and I'll add a `Vec::with_capacity` as default in the builder function to allocate only once.